File: app/src/main/java/com/ignis/android_cleanarchitecture/data/repository/WeatherRepositoryImpl.java
package com.ignis.android_cleanarchitecture.data.repository;

import android.content.Context;

import com.ignis.android_cleanarchitecture.data.api.WeatherApi;
import com.ignis.android_cleanarchitecture.data.api.WeatherRetrofit;
import com.ignis.android_cleanarchitecture.data.dao.WeatherDao;
import com.ignis.android_cleanarchitecture.data.entity.WeatherEntity;
import com.ignis.android_cleanarchitecture.data.entity.mapper.WeatherMapper;
import com.ignis.android_cleanarchitecture.domain.model.weather.WeatherModel;
import com.ignis.android_cleanarchitecture.domain.repository.WeatherRepository;

import io.realm.Realm;
import retrofit2.Retrofit;
import rx.Observable;

/**
 * Weather Repository Implement
 * <p>
 * Created by tamura_k on 2016/05/27.
 */
public class WeatherRepositoryImpl implements WeatherRepository {

    private Context context;

    public WeatherRepositoryImpl(Context context) {
        this.context = context;
    }

    @Override
    public Observable<WeatherModel> fetch(int cityId) {
        Retrofit retrofit = WeatherRetrofit.getInstance();
        WeatherApi weatherApi = retrofit.create(WeatherApi.class);
        return weatherApi.get(cityId)
                .doOnNext(weatherEntity -> {
                    if (exist(cityId)) delete(cityId);
                    weatherEntity.setCityId(cityId);
                    insert(weatherEntity);
                })
                .map(WeatherMapper::convert);
    }

    @Override
    public WeatherModel find(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            WeatherDao weatherDao = new WeatherDao(realm);
            WeatherEntity weatherEntity = weatherDao.find(cityId);
            return WeatherMapper.convert(weatherEntity);
        } finally {
            if (realm != null) realm.close();
        }
    }

    private boolean exist(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            WeatherDao weatherDao = new WeatherDao(realm);
            return weatherDao.exist(cityId);
        } finally {
            if (realm != null) realm.close();
        }
    }

    private void insert(WeatherEntity weatherEntity) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            realm.beginTransaction();
            WeatherDao weatherDao = new WeatherDao(realm);
            weatherDao.insert(weatherEntity);
            realm.commitTransaction();
        } finally {
            if (realm != null) realm.close();
        }
    }

    private void delete(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            realm.beginTransaction();
            WeatherDao weatherDao = new WeatherDao(realm);
            weatherDao.delete(cityId);
            realm.commitTransaction();
        } finally {
            if (realm != null) realm.close();
        }
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/data/repository/weather/WeatherRepositoryImpl.kt
package com.kazakago.cleanarchitecture.data.repository.weather

import android.content.Context
import com.kazakago.cleanarchitecture.data.database.dao.AppDatabase
import com.kazakago.cleanarchitecture.data.database.mapper.weather.WeatherMapper
import com.kazakago.cleanarchitecture.domain.model.weather.Weather
import com.kazakago.cleanarchitecture.domain.repository.weather.WeatherRepository

class WeatherRepositoryImpl(private val context: Context) : WeatherRepository {

    private val database = AppDatabase.create(context)

    override fun find(cityId: String): Weather? {
        val weatherDao = database.weatherDao()
        return weatherDao.findWeather(cityId)?.let {
            WeatherMapper.map(it)
        }
    }

    override fun insert(weather: Weather) {
        val weatherEntity = WeatherMapper.reverse(weather)
        val weatherDao = database.weatherDao()
        weatherDao.insert(weatherEntity)
    }

    override fun delete(weather: Weather) {
        val weatherEntity = WeatherMapper.reverse(weather)
        val weatherDao = database.weatherDao()
        weatherDao.delete(weatherEntity)
    }

}
